### PR TITLE
#184581186: delete a product from user's collection

### PR DESCRIPTION
**Pull request by Rwibutso**
#### What does this PR do?
This PR will delete a specific item from a user collection

#### Description of Task to be completed?
- creating routes for delete a product 
- documenting routes in swagger
- test the endpoint

#### How should this be manually tested?
- Clone this repo git clone https://github.com/atlp-rwanda/e-comm-team-amigos-bn
- Switch your working directory cd e-comm-team-amigos-bn
- Create an .env file and update it with .env-example
- Install all the dependencies by running npm install
- Then run the project by typing npm start

#### Endpoints to be tested:
- /product/delete

#### Any background context you want to provide?
- Have Node Js installed in your local machine
- postman (optional)

#### What are the relevant pivotal tracker/Trello stories?

Story ID: [#184581186](https://www.pivotaltracker.com/story/show/184581186)
#### Screenshots (if appropriate)
Show us the test coverage screenshot and swagger output in your endpoint documentation 
this is the test coverage for the task

Screenshot from 2023-03-22 13-18-55

this is it swagger endpoint

Screenshot from 2023-03-26 10-49-35

### CODE REVIEW

**Review by wilbrord2 — COMMENTED** [review]
Show us the test coverage screenshot and swagger output in your endpoint documentation

File: src/config.js [hunk 1]
```diff
@@ -11,12 +11,12 @@ module.exports = {
...
   },
   test: {
-    use_env_variable: '',
-    username: process.env.DB_USER,
-    password: process.env.DB_PASSWORD,
-    database: process.env.TEST_DB_NAME,
-    host: process.env.DB_HOST,
-    port: process.env.DB_PORT,
+    use_env_variable: 'DATABASE_URL',
+    // username: process.env.DB_USER,
```
Comment (nshutijonathan) [inline src/config.js]: Don't push commented out codes

**Review by nshutijonathan — CHANGES_REQUESTED** [review]
Rename your PR title

File: src/controllers/product.controller.js [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+      expiryDate,
+      ec,
+    } = req.body;
+    const images = req.body.images || [];
+    const existingProduct = await models.Product.findOne({ where: { name } });
+    if (existingProduct) {
+      return res.status(409).json({
+        message: "Product already exists you can update that product instead",
+      });
+    }
```
Comment (Emile-SHUMBUSHO) [inline src/controllers/product.controller.js]: separate validation and controller

File: src/controllers/product.controller.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+export const deleteProduct = async (req, res) => {
+  try {
+    const authHeader = req.headers.authorization;
+    if (!authHeader) {
+      return res.status(401).json({ message: "Authorization header missing" });
+    }
+    const token = authHeader.split(" ")[1];
+    const decodedToken = jwt.verify(token, process.env.SECRET_KEY);
+    const sellerId = decodedToken.userId;
```
Comment (Emile-SHUMBUSHO) [inline src/controllers/product.controller.js]: Better to separate jwt verification and controller

File: src/controllers/product.controller.js [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+      where: { id: productId, userId: sellerId },
+    });
+    if (!product) {
+      return res.status(404).json({ message: "Product not found" });
+    }
+    await product.destroy();
+    res.status(204).end();
+  } catch (error) {
+    res.status(400).json({ error });
+  }
```
Comment (Emile-SHUMBUSHO) [inline src/controllers/product.controller.js]: use status code 400 inside catch block is not defining exact error better to use 500 for internal sever error

**Review by Emile-SHUMBUSHO — CHANGES_REQUESTED** [review]

**Comment by Emile-SHUMBUSHO** [comment]
Rebase is required to resolve a conflicts with develop

**Review by wilbrord2 — COMMENTED** [review]
remove those conflicts

File: src/app.js [hunk 5]
```diff
@@ -15,9 +15,9 @@ dotenv.config();
 const app = express();
 app.use(express.json());
 
-if (process.env.NODE_ENV === "development") {
-  app.use(logger("dev"));
-}
+// if (process.env.NODE_ENV === "development") {
```
Comment (nshutijonathan) [inline src/app.js]: don't push commented out codes

File: src/tests/product.test.js [hunk 6]
```diff
@@ -285,3 +282,122 @@ describe("addProduct function", () => {
     expect(res.body.message).to.equal("Unauthorized access");
   });
 });
+
```
Comment (nshutijonathan) [inline src/tests/product.test.js]: commented out codes should not be pushed

**Review by nshutijonathan — CHANGES_REQUESTED** [review]
LGTM

**Review by wilbrord2 — APPROVED** [review]

**Review by Emile-SHUMBUSHO — APPROVED** [review]

**Review by khevin2 — APPROVED** [review]

**Review by Emile-SHUMBUSHO — COMMENTED** [review]
Cleary show documented screenshot for deleting item not deleting permission

**Comment by Emile-SHUMBUSHO** [comment]
This PR has a conflict to resolve before merging it to develop

**Review by AbdulKhaliq59 — APPROVED** [review]
Resolve conflict and conversation unless LGTM

**Comment by Emile-SHUMBUSHO** [comment]
I think permission/delete/Id was implemented in setting the user role not in the product feature.

File: src/controllers/product.controller.js [hunk 7]
```diff
@@ -265,6 +265,35 @@ export async function updateProduct(req, res) {
...
+
+export const deleteProduct = async (req, res) => {
+    try {
+      const authHeader = req.headers.authorization;
+      if (!authHeader) {
+        return res.status(401).json({ message: "Authorization header missing" });
+      }
+      const token = authHeader.split(" ")[1];
+      const decodedToken = jwt.verify(token, process.env.SECRET_KEY);
+      const sellerId = decodedToken.userId;
```
Comment (Habinezajanvier) [inline src/controllers/product.controller.js]: Don't repeat yourself, use verifyToken middleware to extract user id and use it

**Review by Habinezajanvier — CHANGES_REQUESTED** [review]

**Review by Habinezajanvier — APPROVED** [review]
👍

**Pull request merged**